tests/anoph/conftest.py

```diff
@@ -584,6 +584,318 @@ def simulate_aim_variants(path, contigs, snp_sites, n_sites_low, n_sites_high):
+
+        # Simulate END under variants.
+        random_multipliers = np.random.randint(2, 12, size=len(snp_pos))
+        end = random_multipliers * snp_pos
+        variants_grp.create_dataset(name="END", data=end)
+
+    zarr.consolidate_metadata(zarr_path)
+
+
+def simulate_cnv_discordant_read_calls(zarr_path, metadata_path, contigs, snp_sites):
+    # zarr_path is the output path to the zarr store
+    # metadata_path is the input path for the sample metadata
+    # contigs is the list of contigs, e.g. Ag has ('2R', '3R', 'X')
+    # snp_sites are the data from snp_genotypes/all/sites, inc. {contig}/variants/POS
+    # n_sites will be derived from snp_sites {contig}/variants/POS
+
+    # {release}/cnv/{sample_set}/discordant_read_calls/zarr
+    # - {contig}
+    #   - calldata
+    #     - GT [2D array] [int] [0 or 1 for n_sites for n_samples]
+    #   - samples [1D array] [str for n_samples]
+    #   - variants
+    #      - END [1D array] [int for n_sites]
+    #      - EndBreakpointMethod [1D array] [-1 to 2 for n_sites]
+    #      - ID [1D array] [unique str for n_sites]
+    #      - POS [1D array] [int for n_sites]
+    #      - Region [1D array] [unique str for n_sites]
+    #      - StartBreakpointMethod [1D array] [int] [-1 to 1 for n_sites]
+    # - sample_coverage_variance [1D array] [float] [0 to 0.5 for n_samples]
+    # - sample_is_high_variance [1D array] [bool] [True or False for n_samples]
+    # - samples [1D array] [str for n_samples]
+
+    # Get a random probability for a sample being high variance, between 0 and 1.
+    p_variance = np.random.random()
+
+    # Get a random probability for choosing allele 1, between 0 and 1.
+    p_allele = np.random.random()
+
+    # Open a zarr at the specified path.
+    root = zarr.open(zarr_path, mode="w")
+
+    # Create samples array.
+    df_samples = pd.read_csv(metadata_path)
+    samples = df_samples["sample_id"].values
+    root.create_dataset(name="samples", data=samples, dtype=str)
+
+    # Get the number of samples.
+    n_samples = len(df_samples)
+
+    # Simulate sample_coverage_variance array.
+    sample_coverage_variance = np.random.uniform(low=0, high=0.5, size=n_samples)
+    root.create_dataset(name="sample_coverage_variance", data=sample_coverage_variance)
+
+    # Simulate sample_is_high_variance array.
+    sample_is_high_variance = np.random.choice(
+        [False, True], size=n_samples, p=[1 - p_variance, p_variance]
+    )
+    root.create_dataset(name="sample_is_high_variance", data=sample_is_high_variance)
+
+    for contig in contigs:
+        # Create the contig group.
+        contig_grp = root.require_group(contig)
+
+        # Create the calldata group for this contig.
+        calldata_grp = contig_grp.require_group("calldata")
+
+        # Get the SNP positions for this contig.
+        snp_pos = snp_sites[f"{contig}/variants/POS"][:]
+
+        # Get the number of sites (SNP positions) for this contig.
+        n_sites = snp_pos.shape[0]
```

Comment:
Again these data are independent of SNPs. They are similar to the coverage calls, in that each CNV variant has a start coordinate and an end coordinate. So you could take a similar approach to simulating these.

Of course there are much fewer variants in the discordant reads data than in the coverage calls.

Reply:
Thanks. Have done.
The catch in this case is that the number of variants need to be consistent for each contig across all sample sets, because the data get concatenated that way.